db/database.go

```diff
@@ -252,6 +252,16 @@ type DBOnlineCallback func(dbContext *DatabaseContext)
 
 // Creates a new DatabaseContext on a bucket. The bucket will be closed when this context closes.
 func NewDatabaseContext(dbName string, bucket base.Bucket, autoImport bool, options DatabaseContextOptions) (*DatabaseContext, error) {
+	var returnedError error
+	cleanupFunctions := make([]func(), 0)
+
+	defer func() {
+		if returnedError != nil {
+			for _, cleanupFunc := range cleanupFunctions {
+				cleanupFunc()
+			}
+		}
+	}()
```

Comment:
Turns out you can do plain returns without assignment and still have a deferred function read the value.

https://play.golang.org/p/DhWkH8CAoPf

Give this a go and see how much it can clean it up to avoid the "returnedError" assignments, I think it will help a lot.